assets/js/googlesitekit/widgets/util/combine-widgets.js

```diff
@@ -0,0 +1,119 @@
+ * @param {Array.<Array>}  layout.classNames   List of class name arrays for
+ *                                             each widget.
+ * @param {Array.<number>} layout.columnWidths List of column widths for each
+ *                                             widget.
+ * @param {Array.<number>} layout.rowIndexes   List of row indexes for each
+ *                                             widget.
+ * @return {Object} Object with `gridClassNames` and `overrideComponents`
+ *                  properties, both of which are a list with one item for each
+ *                  widget. Every `gridClassNames` entry is an array of class
+ *                  names, every `overrideComponents` entry is either an object
+ *                  with `Component` and `metadata`, or `null` (similar to
+ *                  the `widgetStates` parameter).
+ */
+export function combineWidgets( activeWidgets, widgetStates, {
+	classNames,
+	columnWidths,
+	rowIndexes,
+} ) {
+	const gridClassNames = [ ...classNames ];
+	const overrideComponents = [];
+
+	let currentState = null;
+	let currentRowIndex = -1;
+	let columnWidthsBuffer = [];
+
+	activeWidgets.forEach( ( widget, i ) => {
+		overrideComponents.push( null );
+
+		// Hide any widgets that have `null` as class names (which happens
+		// when the widget renders `null`).
+		if ( gridClassNames[ i ] === null ) {
+			gridClassNames[ i ] = [ HIDDEN_CLASS ];
+			return;
+		}
+
+		currentState = widgetStates[ widget.slug ];
+		currentRowIndex = rowIndexes[ i ];
+
+		// If the current widget has a special state...
+		if ( currentState ) {
+			if ( stateAndRowMatch( currentState, widgetStates[ activeWidgets[ i + 1 ]?.slug ], currentRowIndex, rowIndexes[ i + 1 ] ) ) {
+				// If the current widget state and row index match the next
+				// state and row index, hide the widget entirely. Only the last
+				// similar instance will be rendered in this case.
+				gridClassNames[ i ] = [ HIDDEN_CLASS ];
+				columnWidthsBuffer.push( columnWidths[ i ] );
+			} else if ( columnWidthsBuffer.length > 0 ) {
+				// If the state and row index do not match the next ones and
+				// there are already similar instances (from previous
+				// iterations), this is the last similar instance, so the
+				// combined version will need to be displayed instead of the
+				// widget. The combined version will use the common Component
+				// and pass all common metadata as props.
+				columnWidthsBuffer.push( columnWidths[ i ] );
+
+				// Get total (desktop) column width and use corresponding grid
+				// classes. For tablet and phone, the component should span the
+				// full width as by definition it is at least a "half" widget
+				// wide (which has that behavior).
+				const combinedColumnWidth = columnWidthsBuffer.reduce( ( sum, columnWidth ) => sum + columnWidth, 0 );
+
+				gridClassNames[ i ] = [
+					'mdc-layout-grid__cell',
+					`mdc-layout-grid__cell--span-${ combinedColumnWidth }`,
+				];
```

Comment:
That's a good point. However, the concept of passing around class names like this has already been there before this PR to compose the widget area layout. Changing it in this PR would make it even more complex and potentially result in a delay. I'll create a follow-up issue to refactor that out, I agree it's a simplification worth adding. Since these functions and components are only used within Site Kit, changing it will be okay even when the Widgets API is production-ready.

Reply:
#2671